File: Python/treehopper/libraries/sensors/pressure/Bme280.py
from math import log
from typing import List

from treehopper.api import I2C
from treehopper.libraries import SMBusDevice
from treehopper.libraries.register_manager import sign_extend
from treehopper.libraries.sensors.humidity.humidity_sensor import HumiditySensor
from treehopper.libraries.sensors.pressure import Bmp280
from treehopper.libraries.sensors.pressure.bmp280_registers import Oversamplings


class Bme280(Bmp280, HumiditySensor):
    @staticmethod
    def probe(i2c: I2C) -> List['Bme280']:
        devs = []  # type: List['Bme280']
        try:
            dev = SMBusDevice(0x76, i2c, 100)
            who_am_i = dev.read_byte_data(0xD0)
            if who_am_i == 0x60:
                devs.append(Bme280(i2c, False))
        except RuntimeError:
            pass

        try:
            dev = SMBusDevice(0x76, i2c, 100)
            who_am_i = dev.read_byte_data(0xD0)
            if who_am_i == 0x60:
                devs.append(Bme280(i2c, True))
        except RuntimeError:
            pass

        return devs

    def __init__(self, i2c: I2C, sdo=False, rate=100.0):
        super().__init__(i2c, sdo, rate)
        self._registers.readRange(self._registers.h2, self._registers.h6)

        self._h4 = sign_extend(self._registers.h4.value << 4 | self._registers.h4h5.h4Low, 12)
        self._h5 = sign_extend(self._registers.h5.value << 4 | self._registers.h4h5.h5Low, 12)

        self._registers.ctrlHumidity.oversampling = Oversamplings.Oversampling_x16
        self._registers.ctrlHumidity.write()
        self._registers.ctrlMeasure.write()

    def update(self):
        super().update()

        var_h = self.t_fine - 76800.0
        var_h = (self._registers.humidity.value - (self._h4 * 64.0 + self._h5 / 16384.0 * var_h)) * \
                self._registers.h2.value / 65536.0 * (1.0 + self._registers.h6.value / 67108864.0 * var_h *
                                                      (1.0 + self._registers.h3.value / 67108864.0 * var_h))

        var_h = var_h * (1.0 - self._registers.h1.value * var_h / 524288.0)

        if var_h > 100.0:
            var_h = 100.0
        elif var_h < 0.0:
            var_h = 0.0

        self._relative_humidity = var_h

File: NET/API/Treehopper.Libraries/Sensors/Pressure/Bme280.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Treehopper.Libraries.Sensors.Humidity;

namespace Treehopper.Libraries.Sensors.Pressure
{
    /// <summary>
    ///     Bosch BME280 barometric pressure, temperature, and humidity sensor
    /// </summary>
    [Supports("Bosch", "BME280")]
    public class Bme280 : Bmp280, IHumiditySensor
    {
        /// <summary>
        /// Probes the specified I2C bus to discover any BME280 sensors attached.
        /// </summary>
        /// <param name="i2c">The bus to probe.</param>
        /// <param name="rate">The rate, in kHz, to use.</param>
        /// <returns>An awaitable task that completes with a list of BME280 sensors found.</returns>
        public static async Task<IList<Bme280>> ProbeAsync(I2C i2c, int rate=100)
        {
            var deviceList = new List<Bme280>();
            try
            {
                var dev = new SMBusDevice(0x76, i2c, 100);
                var whoAmI = await dev.ReadByteDataAsync(0xD0).ConfigureAwait(false);
                if (whoAmI == 0x60)
                    deviceList.Add(new Bme280(i2c, false));
            }
            catch (Exception ex) { }

            try
            {
                var dev = new SMBusDevice(0x77, i2c, 100);
                var whoAmI = await dev.ReadByteDataAsync(0xD0).ConfigureAwait(false);
                if (whoAmI == 0x60)
                    deviceList.Add(new Bme280(i2c, true));
            }
            catch (Exception ex) { }
            return deviceList;
        }

        private double humidity;
        private short h4, h5;
        /// <summary>
        ///     Construct a BMP280 hooked up to the i2C bus
        /// </summary>
        /// <param name="i2c">the i2C bus to use</param>
        /// <param name="sdo">the state of the SDO pin, which sets the address</param>
        public Bme280(I2C i2c, bool sdo = false) : base(i2c, sdo)
        {
            Task.Run(() => registers.readRange(registers.h2, registers.h6)).Wait();

            // RegisterGenerator doesn't get the endianness right on the h4/h5 12-bit values, so manually create them:
            h4 = (short)((short)((registers.h4.value << 4 | registers.h4h5.h4Low) << 4) >> 4);
            h5 = (short)((short)((registers.h5.value << 4 | registers.h4h5.h5Low) << 4) >> 4);

            registers.ctrlHumidity.setOversampling(Oversamplings.Oversampling_x16);
            Task.Run(registers.ctrlHumidity.writeAsync).Wait();
            Task.Run(registers.ctrlMeasure.writeAsync).Wait();
        }

        /// <summary>
        ///     The relative humidity
        /// </summary>
        public double RelativeHumidity
        {
            get
            {
                if (AutoUpdateWhenPropertyRead) UpdateAsync().Wait();
                return humidity;
            }
        }

        /// <summary>
        /// Requests a reading from the sensor and updates its data properties with the gathered values.
        /// </summary>
        /// <returns>An awaitable Task</returns>
        /// <remarks>
        /// Note that when #AutoUpdateWhenPropertyRead is `true` (which it is, by default), this method is implicitly 
        /// called when any sensor data property is read from --- there's no need to call this method unless you set
        /// AutoUpdateWhenPropertyRead to `false`.
        /// 
        /// Unless otherwise noted, this method updates all sensor data simultaneously, which can often lead to more efficient
        /// bus usage (as well as reducing USB chattiness).
        /// </remarks>
        public override async Task UpdateAsync()
        {
            // first the BMP280 stuff
            await base.UpdateAsync().ConfigureAwait(false);

            // now the BME stuff
            double var_H;

            var_H = tFine - 76800.0;
            var_H = (registers.humidity.value - (h4 * 64.0 + h5 / 16384.0 * var_H)) *
                registers.h2.value / 65536.0 * 
                (1.0 + registers.h6.value / 67108864.0 * var_H *
                (1.0 + registers.h3.value / 67108864.0 * var_H));

            var_H = var_H * (1.0 - registers.h1.value * var_H / 524288.0);

            if (var_H > 100.0)
                var_H = 100.0;
            else if (var_H < 0.0)
                var_H = 0.0;

            humidity = var_H;
        }
    }
}